### PR TITLE
Feat/admin/webhooks

### PR DESCRIPTION
**Pull request by ryanhopperlowe**
Screenshot 2024-11-20 at 11 14 18 AM
Screenshot 2024-11-20 at 11 14 27 AM

### CODE REVIEW

**Comment by ryanhopperlowe** [comment]
Sorry for the long PR. I'll do a better job of splitting things up going forward

File: ui/admin/app/components/webhooks/WebhookConfirmation.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    return (
+        <Dialog open>
+            <DialogContent className="max-w-[700px]" hideCloseButton>
+                <DialogHeader>
+                    <DialogTitle>Webhook Saved</DialogTitle>
+                </DialogHeader>
+
+                <DialogDescription>
+                    Your webhook has been saved in Otto. Make sure to copy
+                    payload url and secret to your webhook provider.
```
Comment (g-linville) [inline ui/admin/app/components/webhooks/WebhookConfirmation.tsx]: nit:

```suggestion
                    Your webhook has been saved in Otto8. Make sure to copy
                    the payload URL and secret to your webhook provider.
```

File: ui/admin/app/components/webhooks/WebhookForm.tsx [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+            <FormItem>
+                <FormLabel>Type</FormLabel>
+                <Select value="Github" disabled>
+                    <SelectTrigger>
+                        <SelectValue />
+                    </SelectTrigger>
+
+                    <SelectContent>
+                        <SelectItem value="Github">Github</SelectItem>
```
Comment (g-linville) [inline ui/admin/app/components/webhooks/WebhookForm.tsx]: nit:

```suggestion
                <Select value="GitHub" disabled>
                    <SelectTrigger>
                        <SelectValue />
                    </SelectTrigger>

                    <SelectContent>
                        <SelectItem value="GitHub">GitHub</SelectItem>
```

File: ui/admin/app/components/webhooks/WebhookForm.tsx [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+                label="Secret"
+                description="This secret should match the secret you provide to GitHub."
+                placeholder={hasSecret ? "(unchanged)" : ""}
+            />
+
+            <ControlledInput
+                control={form.control}
+                name="token"
+                label="Token (optional)"
+                description="Optionally provide a token to add an extra layer of security."
```
Comment (g-linville) [inline ui/admin/app/components/webhooks/WebhookForm.tsx]: I think we should either add more information here or link to docs (if they exist) to explain what this does. Whatever token is set will have to be provided by the webhook provider (i.e. GitHub) in a particular query param or header.

File: ui/admin/app/components/webhooks/WebhookFormContext.tsx [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        });
+    });
+
+    return (
+        <Form {...form}>
+            <Context.Provider
+                value={{
+                    error: updateWebhook.error || createWebhook.error,
+                    isEdit: !!webhookId,
+                    hasSecret: !!webhookId,
```
Comment (g-linville) [inline ui/admin/app/components/webhooks/WebhookFormContext.tsx]: I'm confused about how the webhook ID would be related to whether or not the webhook has a secret set...

File: ui/admin/app/components/workflow/steps/While.tsx [hunk 5]
```diff
@@ -51,7 +51,7 @@ export function WhileComponent({
 
     return (
         <div className={cn("border rounded-md", className)}>
-            <ButtonDiv
+            <ClickableDiv
```
Comment (StrongMonkey) [inline ui/admin/app/components/workflow/steps/While.tsx:54]: Oh cool, I should have used that

File: ui/admin/app/components/webhooks/WebhookForm.tsx [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+
+        return workflows.map((workflow) => (
+            <SelectItem key={workflow.id} value={workflow.id}>
+                {workflow.name}
+            </SelectItem>
+        ));
+    }
+}
+
+const GithubHeaderOptions = [
```
Comment (StrongMonkey) [inline ui/admin/app/components/webhooks/WebhookForm.tsx:229]: Do we allow any header? these headers seems to be specific to github

File: ui/admin/app/components/webhooks/WebhookForm.tsx [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    "X-GitHub-Hook-ID",
+    "X-GitHub-Event",
+    "X-GitHub-Delivery",
+    "X-Hub-Signature",
+    "X-Hub-Signature-Ed25519",
+    "User-Agent",
+    "X-GitHub-Hook-Installation-Target-Type",
+    "X-GitHub-Hook-Installation-Target-ID",
```
Comment (thedadams) [inline ui/admin/app/components/webhooks/WebhookForm.tsx]: Glad to see this isn't a dropdown and is just a list of suggestions. Can we remove the signature headers? They don't seem useful here and the user can type them in if they really want it.

**Review by thedadams — COMMENTED** [review]
I noticed that the webhook create and edit pages take up the full-screen width. It looks a little weird on my larger monitor. Does it make sense to shrink the width down?

File: ui/admin/app/components/webhooks/WebhookFormContext.tsx (hunk 4 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookFormContext.tsx]: You can't create a webhook without a secret right? I was thinking that anytime a webhook exists in db (aka has an id), it must have a secret.

That being said it probably makes more sense to check for a secret hash instead

File: ui/admin/app/components/workflow/steps/While.tsx (hunk 5 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/workflow/steps/While.tsx:54]: Yeah I got annoyed that shadcn doesn't provide a camouflaged button

File: ui/admin/app/components/webhooks/WebhookForm.tsx (hunk 6 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookForm.tsx:229]: This form will be adapted in the future to switch between various providers. For now it's specifically for github

File: ui/admin/app/components/webhooks/WebhookForm.tsx (hunk 3 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookForm.tsx]: Ah, I mistakenly assumed the token would be provided in the returned `link` meta. I'll update this

File: ui/admin/app/components/webhooks/WebhookFormContext.tsx (hunk 4 above)
Comment (thedadams) [reply ui/admin/app/components/webhooks/WebhookFormContext.tsx]: A webhook does not have to have a secret.

File: ui/admin/app/components/webhooks/WebhookForm.tsx (hunk 7 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookForm.tsx]: done

File: ui/admin/app/components/webhooks/WebhookFormContext.tsx (hunk 4 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookFormContext.tsx]: discussed with Donnie offline. This has been updated

File: ui/admin/app/components/webhooks/WebhookForm.tsx (hunk 3 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookForm.tsx]: I discussed this with Will and he decided to hide the token feature entirely until we can talk it out more with Bill/Donnie

File: ui/admin/app/components/webhooks/WebhookForm.tsx (hunk 2 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookForm.tsx]: done

File: ui/admin/app/components/webhooks/WebhookConfirmation.tsx (hunk 1 above)
Comment (ryanhopperlowe) [reply ui/admin/app/components/webhooks/WebhookConfirmation.tsx]: done

**Comment by thedadams** [comment]
A few thoughts here:
1. Let's remove alias until we get feedback from Craig on it.
2. I am not able to click the "No Secret" checkbox. Also, not sure what it is used for. The user isn't required to enter a secret.
3. `*` shouldn't be the default value for headers. Maybe add some detail text to explain that `*` can be used to include all headers.

Screenshot 2024-11-22 at 06 39 10

**Review by njhale — APPROVED** [review]

**Review by thedadams — APPROVED** [review]

**Pull request merged**